parlai/utils/torch.py

```diff
@@ -0,0 +1,223 @@
+#!/usr/bin/env python3
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+Utility methods for dealing with torch code.
+"""
+
+from typing import Union, Optional, Tuple, Any, List, Sized
+
+try:
+    import torch
+except ImportError:
+    raise ImportError('Parlai requires pytorch. Go to http://pytorch.org to install.')
+
+import torch.optim
+
+"""Near infinity, useful as a large penalty for scoring when inf is bad."""
+NEAR_INF = 1e20
+NEAR_INF_FP16 = 65504
+
+
+def neginf(dtype: torch.dtype) -> float:
+    """
+    Return a representable finite number near -inf for a dtype.
+    """
+    if dtype is torch.float16:
+        return -NEAR_INF_FP16
+    else:
+        return -NEAR_INF
+
+
+def padded_tensor(
+    items: List[Union[List[int], torch.LongTensor]],
+    pad_idx: int = 0,
+    use_cuda: bool = False,
+    left_padded: bool = False,
+    max_len: Optional[int] = None,
+    fp16friendly: bool = False,
+) -> Tuple[torch.LongTensor, List[int]]:
+    """
+    Create a right-padded matrix from an uneven list of lists.
+
+    Returns (padded, lengths), where padded is the padded matrix, and lengths
+    is a list containing the lengths of each row.
+
+    Matrix is right-padded (filled to the right) by default, but can be
+    left padded if the flag is set to True.
+
+    Matrix can also be placed on cuda automatically.
+
+    :param list[iter[int]] items: List of items
+    :param bool sort: If True, orders by the length
+    :param int pad_idx: the value to use for padding
+    :param bool use_cuda: if true, places `padded` on GPU
+    :param bool left_padded:
+    :param int max_len: if None, the max length is the maximum item length
+    :param bool fp16friendly: if True, pads the time dimension to be a multiple of 8.
+
+    :returns: (padded, lengths) tuple
+    :rtype: (Tensor[int64], list[int])
+    """
+
+    # number of items
+    n = len(items)
+    # length of each item
+    lens: List[int] = [len(item) for item in items]  # type: ignore
```

Comment:
why ignore this type?

Reply:
torch.Tensor is not of type Sizeable, so this throws an error. (Generally len(tensor) does). I'd like to make a patch for that.